Jss modify tags

jschneid opened the pull request:
Closes #14

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -21,6 +23,22 @@ def create
 
+  # GET /images/1/edit
+  def edit
+    @image = Image.find(params[:id])
```
there could be a case when image id is invalid. In rails `find` would throw an exception if it's not existed, so I would recommend using `find_by`, and deal with the non-existent case.

jschneid replied:
Ok! Updated in 1202d73.

wilsonyqm commented on test/controllers/images_controller_test.rb:
```diff
@@ -103,6 +103,11 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
 
+  test 'attempt to load the view image page with a nonexistent image ID' do
+    get image_path('invalid-id-value')
+    assert_response :not_found
```
we probably need to do more than just a HTTP response.

jschneid replied:
Added validation that the correct error message is being displayed in 97416a5. (Anything else that would be helpful to assert in this case?)

wilsonyqm replied:
Yes, the tags could not be always short. Imagine you have a very long tag, would your app able to handle it?

jschneid replied:
The model, image.rb, does currently set a maximum tag length of 20 characters. (https://github.com/AppFolioOnboarding/image-sharer-jschneid/blob/1c9118d083c012eae5ebad520e7870c6185859ca/app/models/image.rb#L12) 

Can you help me better understand your comment here -- I'm not sure what the `get` method on the image controller has to do with tag lengths, or with other image properties beyond the ID?  What additional test coverage would you suggest be added (if any)?

wilsonyqm replied:
There is a bug happened, when I tried to type in a very long tag in your app, I tried it on heroku so I don't know what the error is, but I think we need to handle it. The validation looks good, but somehow it is not working, would recommend to check the bug first.

jschneid replied:
You're right! There was invalid SQL being generated due to the bug fixed by 45855cd. Thanks for catching that! Should be working on Heroku now!

jschneid replied:
I'm not sure why, but in my local environment, and in the unit tests, the wrong(?) syntax that had been fixed in https://github.com/AppFolioOnboarding/image-sharer-jschneid/commit/45855cdd1381f46e4fa31c5c7e5611c6e3ec8704 was actually working ok -- it was just in Heroku that the problem was manifesting.

When I tried changing it to syntax `find_by(id: ...)`, I was still having issues avoiding false positive test passes.

I decided to deal with this by changing all of those `find_by(id: ...)` calls to just `find(...)` instead. Correspondingly, I needed to error handle by rescuing `ActiveRecord::RecordNotFound` instead of checking for `nil?`. This change is in 0c6982d.

wilsonyqm commented on config/routes.rb:
```diff
@@ -1,7 +1,7 @@
   root to: 'images#index'
 
-  resources :images, only: %i[show new create destroy]
+  resources :images, only: %i[show new create edit update destroy]
```
Since you have everything,  you probably don't need the `only` here.

jschneid replied:
Oh! Good catch! I didn't even think about how, now that we've added `edit` and `update`, we've added everything there is to add for this! Fixed in 1c9118d.

wilsonyqm commented on test/controllers/images_controller_test.rb:
```diff
@@ -166,4 +171,78 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
+
+    # === Assert ===
+
+    # Verify we're still on the Edit page, that the error message displayed, and that the
```
I think having comments is not necessary, can just leave a blank line between each steps.

jschneid replied:
Are you referring to the Arrange/Act/Assert steps (and their corresponding "section header" comments), or to these specific comments in the Assert section?

For the section comments, we found that they were pretty useful in the RentLinx project, but test code in that project was often pretty heavyweight, such that the "section headers" actually were pretty helpful in making sense of the code. In these more lightweight Rails tests, though, I agree that they are actually a little more "noisy" than helpful.  I actually did omit their use in the other controller test files in this Ropes project.

For the comments on this specific Assert section, I agree that the code on its own is pretty straightforward and doesn't necessarily merit a comment, but I kind of like leaving the comment there so anyone reading the code can tell that the developer's intent here does match what's actually implemented (instead of having to speculate about that, in the event that something looks wrong or missing in the code).  Is it ok with you if I leave it in place?

wilsonyqm replied:
Definitely fine to keep that. Usually I prefer to leave a comment only if it's really hard to understand code or there is some special cases. Anyway, good to know.

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -1,4 +1,6 @@
 class ImagesController < ApplicationController
+  rescue_from ActiveRecord::RecordNotFound, with: :render_image_not_found
```
RecordNotFound could be thrown from somewhere else except the image, so I recommend not do in this way. https://github.com/AppFolioOnboarding/image-sharer-jschneid/pull/33/files#r152119609

jschneid replied:
Ok! Updated in 1202d73.

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -43,7 +62,16 @@ def destroy
+
+  def render_image_not_found
+    flash[:danger] = 'Image not found.'
+    render :not_found, status: :not_found
```
You don't need to render `not_found` page, it's better to stay at the same page, and show the error under the form fields

jschneid replied:
As it stands currently, I think the "not found" page will be displayed in the following situations:
- Someone takes a valid "view image" page URL, e.g. http://localhost:3000/images/5, and manually edits the URL to have an invalid image ID, e.g. http://localhost:3000/images/999999
- Someone tries to view or edit an image, but that image has been deleted -- perhaps by the same user in a different browser tab, or by another user.

Since you mentioned "form fields," were you thinking of having the user stay on the "edit image" form? I'm not sure it would be helpful to the user to keep showing them the edit image form in that case, since the image the page is trying to edit doesn't exist (since it has either been deleted, or the user manually edited the URL to have an invalid image ID), and there are no possible values the user can supply in the form fields to correct that problem?

wilsonyqm replied:
That makes sense, thanks for the explanation, I think redirect to the index page would be better and show a flash message like: `The image is not available`, the not found page is not necessary, we don't use that in property app. How do you think?

jschneid replied:
That's a good suggestion! Made that change in 0c6982d.

wilsonyqm commented on test/controllers/images_controller_test.rb:
```diff
@@ -103,6 +103,12 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test 'attempt to load the view image page with a nonexistent image ID' do
+    get image_path('invalid-id-value')
+    assert_response :not_found
+    assert_select 'p', 'Image not found.'
```
For the error message class, can you specify it, so that it would be clearer than `p`

jschneid replied:
Good suggestion! Fixed in dc0a28b.

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -2,6 +2,8 @@ class ImagesController < ApplicationController
   # GET /images/1
   def show
     @image = Image.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
```
Can you try to use `find_by` instead of `find` and `rescue` here, usually we don't use `rescue` to catch non-existed record.

jschneid replied:
Fixed in 7900238; full commentary here: https://github.com/AppFolioOnboarding/image-sharer-jschneid/pull/33#issuecomment-347210436.

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -21,29 +22,55 @@ def create
+    else
+      render :edit, status: :bad_request
+    end
+  rescue ActiveRecord::RecordNotFound
```
https://github.com/AppFolioOnboarding/image-sharer-jschneid/pull/33#discussion_r152691818

jschneid replied:
Fixed in 7900238; full commentary here: https://github.com/AppFolioOnboarding/image-sharer-jschneid/pull/33#issuecomment-347210436.

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -21,29 +22,55 @@ def create
+  # GET /images/1/edit
+  def edit
+    @image = Image.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
```
https://github.com/AppFolioOnboarding/image-sharer-jschneid/pull/33#discussion_r152691818

jschneid replied:
Fixed in 7900238; full commentary here: https://github.com/AppFolioOnboarding/image-sharer-jschneid/pull/33#issuecomment-347210436.

wilsonyqm commented:
I am not sure why you remove all the `find` into `find_by`, can you point out what's making it not working?

jschneid commented:
> I am not sure why you remove all the `find` into `find_by`, can you point out what's making it not working?

I don't 100% remember, to be honest! I think it was probably one or more of:
- I may have learned about `find` after originally writing the code with `find_by`, and the `find` syntax seemed more succinct (with there being no need to explicitly identify the `id` parameter by name), so I changed it?
- I had some bug in the code related to the use of `find_by` that I couldn't figure out, but it did start working when I changed it to `find`, and I didn't see any disadvantage to `find`, so I just left it?

In any event, though, I think I've now got things working well with `find_by`, so I switched back to use of that, and replaced the `rescue` error handling with `nil?` checks. In 7900238.
